compat/mingw.c

```diff
@@ -1287,35 +1287,61 @@ struct tm *localtime_r(const time_t *timep, struct tm *result)
 	return NULL;
 }
 
-char *mingw_strbuf_realpath(struct strbuf *resolved, const char *path)
+static int realpath_existing_file(struct strbuf *resolved, const char *path, int path_len)
 {
 	wchar_t wpath[MAX_PATH];
 	HANDLE h;
 	DWORD ret;
 	int len;
 
-	if (xutftowcs_path(wpath, path) < 0)
-		return NULL;
+	if (xutftowcs_path_ex(wpath, path, MAX_PATH, path_len, MAX_PATH, 0) < 0)
+		return -1;
 
 	h = CreateFileW(wpath, 0,
 			FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, NULL,
 			OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, NULL);
 	if (h == INVALID_HANDLE_VALUE)
-		return NULL;
+		return -1;
 
 	ret = GetFinalPathNameByHandleW(h, wpath, ARRAY_SIZE(wpath), 0);
 	CloseHandle(h);
 	if (!ret || ret >= ARRAY_SIZE(wpath))
-		return NULL;
+		return -1;
 
 	len = wcslen(wpath) * 3;
 	strbuf_grow(resolved, len);
 	len = xwcstoutf(resolved->buf, normalize_ntpath(wpath), len);
 	if (len < 0)
-		return NULL;
+		return -1;
 	resolved->len = len;
-	return resolved->buf;
+	return 0;
+}
+
+char *mingw_strbuf_realpath(struct strbuf *resolved, const char *path)
+{
+	char *last_component;
+
+	if (realpath_existing_file(resolved, path, -1) >= 0)
+		return resolved->buf;
+
+	// strbuf_realpath() allows last path component to not exist.
+	// Resolving full path failed, now it's time to try without last component.
 
+	// Retry on all errors to better match behavior of `strbuf_realpath()`.
+	// For example, for no-access last directory, `strbuf_realpath()` ignores
+	// that (because it doesn't try to `CreateFile()`) and still succeeds.
```

Comment:
Good comment, but in Git's source code it cannot be C++ style, it has to be `/* ... */`.

Reply:
Good catch!